### PR TITLE
feat: Common Panel Support(Integrate weekly panel) (close #38)

### LINKED ISSUE #38: feat: Common Panel Support(Integrate Milestone, Task, AllDay View and Panel)
## Version

v0.9.0

## Development Environment

Mac
Chrome

## Current Behavior

There are duplicate source code for weekly panel.
`milestone.js`, `taskview.js`, `allday.js` has common functionality but has own source code.

```js
// Write example code
```

## Expected Behavior

Generalize weekly panel for `milestone`, `task`, `allday`.

### PR DESCRIPTION
**Pull request by dongsik-yoo**
### Please check if the PR fulfills these requirements
- [x] It's submitted to right branch according to our branching model
- [x] It's right issue type on title
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added/updated (for bug fixes/features)
- [x] It does not introduce a breaking change or has description for the breaking change

### Description
Integrate milestone, taskview, allday view as DayGrid view.
* dayGrid.js, dayGridSchedule.js
* dayGrid.hbs, dayGridSchedule.hbs
* dayGrid.styl
* handler/daygrid
 * click.js, core.js, creation.js, creationGuide.js, move.js, moveGuide.js, resize.js

#### Panel Information
Handler array will be taken by option.

```js
var DEFAULT_PANELS = [
    {
        name: 'milestone',
        type: 'daygrid',
        minHeight: 20,
        maxHeight: 80,
        showExpandableButton: false,
        maxExpandableHeight: 220,
        handlers: ['click'],
        show: true
    },
    {
        name: 'task',
        type: 'daygrid',
        minHeight: 40,
        maxHeight: 120,
        showExpandableButton: false,
        maxExpandableHeight: 220,
        handlers: ['click', 'move', 'resize'],
        show: true
    },
    {
        name: 'allday',
        type: 'daygrid',
        minHeight: 20,
        maxHeight: 80,
        showExpandableButton: true,
        maxExpandableHeight: 270,
        handlers: ['click', 'creation', 'move', 'resize'],
        show: true
    },
    {
        name: 'time',
        type: 'timegrid',
        autoHeight: true,
        handlers: ['click', 'creation', 'move', 'resize'],
        show: true
    }
];
```

---
Thank you for your contribution to TOAST UI product. 🎉 😘 ✨

### CODE REVIEW

**Comment by dongsik-yoo** [comment]
@youjung-hong local server test http://10.78.65.68:8080/examples/example01-basic.html

File: src/js/controller/viewMixin/week.js [hunk 1]
```diff
@@ -219,12 +219,14 @@ var Week = {
      * @param {Date} start start date.
      * @param {Date} end end date.
      * @param {function[]} [andFilters] - optional filters to applying search query
+     * @param {Array.<Objecy[]>} panels - schedule panels like 'milestone', 'task', 'allday', 'time'
```
Comment (youjung-hong) [inline src/js/controller/viewMixin/week.js]: 1) `panels` type is `Object[]`.
2) Is there any case that having `panels` but no `andFilter`?

Comment (dongsik-yoo) [reply src/js/controller/viewMixin/week.js]: 1. Thanks a lot.
2. I will change parameter sequence. So optional parameter is last.
```js
/**
 * Populate schedules in date range.
 * @this Base
 * @param {Date} start start date.
 * @param {Date} end end date.
 * @param {Array.<object[]>} panels - schedule panels like 'milestone', 'task', 'allday', 'time'
 * @param {function[]} [andFilters] - optional filters to applying search query
 * @returns {object} schedules grouped by dates.
 */
findByDateRange: function(start, end, panels, andFilters) {
}
```

File: src/js/common/vpanel.js [hunk 2]
```diff
@@ -64,6 +64,30 @@ VPanel.prototype.isSplitter = function() {
     return this.options.isSplitter;
 };
 
+/**
+ * set max height of panel
+ * @param {number} maxHeight - maxHeight
+ */
+VPanel.prototype.setMaxHeight = function(maxHeight) {
+    this.options.maxHeight = maxHeight;
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/common/vpanel.js]: As you know we have some view structures, VLayout & VPanel and View.
Originally VPanel's height was set by view's `afterRender` event.
This means those are related with each other .

```js
milestoneView.on('afterRender', function(viewModel) {	
    vLayout.getPanelByName('Milestone').setHeight(null, viewModel.height);	
});
```

I just moved it on DayGrid View. And the common handler can be used in DayGrid View also.

In conclusion, VPanel's properties including `maxHeight`, `HeightForcedSet`, `Height` can be set by View to fit height of View.

File: src/js/handler/daygrid/moveGuide.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        range = dragStartEventData.range,
+        baseWidthPercent = (100 / datesInRange),
+        originScheduleStarts = datetime.start(model.start),
+        originScheduleEnds = datetime.end(model.end),
+        renderStartDate = datetime.start(range[0]),
+        renderEndDate = datetime.end(range[range.length - 1]),
+        fromLeft = (new TZDate(originScheduleStarts.getTime() -
+            renderStartDate.getTime())) / datetime.MILLISECONDS_PER_DAY | 0,
+        fromRight = (new TZDate(originScheduleEnds.getTime() -
+            renderEndDate.getTime())) / datetime.MILLISECONDS_PER_DAY | 0;
```
Comment (youjung-hong) [inline src/js/handler/daygrid/moveGuide.js]: `I` is bitwise operators. - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_Operators

File: src/js/handler/daygrid/resizeGuide.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+ * Return function that calculate guide element's new width percentage value.
+ * @param {object} dragStartEventData - dragstart schedule data.
+ * @returns {function} return function that calculate guide element new width percentage.
+ */
+DayGridResizeGuide.prototype.getGuideElementWidthFunc = function(dragStartEventData) {
+    var model = dragStartEventData.model,
+        viewOptions = this.resizeHandler.view.options,
+        fromLeft = (new TZDate(
+            model.start.getTime() - datetime.parse(viewOptions.renderStartDate)
+        )) / datetime.MILLISECONDS_PER_DAY | 0,
```
Comment (youjung-hong) [inline src/js/handler/daygrid/resizeGuide.js]: please change to `||`

File: src/js/handler/daygrid/resizeGuide.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        var width = 0;
+        var i = 0;
+        var length = grids.length;
+        width += grids[fromLeft] ? grids[fromLeft].width : 0;
+
+        for (; i < length; i += 1) {
+            if (i > fromLeft && i <= xIndex) {
+                width += grids[i] ? grids[i].width : 0;
+            }
+        }
```
Comment (youjung-hong) [inline src/js/handler/daygrid/resizeGuide.js]: this statement is more easy to ready => `for(i = fromLeft; i < length && i < xIndex; i += 1)`

File: src/js/factory/weekView.js [hunk 6]
```diff
@@ -14,60 +14,80 @@ var Week = require('../view/week/week');
...
+    {
+        name: 'allday',
+        type: 'daygrid',
         minHeight: 20,
-        height: 80,
         maxHeight: 80,
-        show: true,
-        maxExpandCount: 10
+        showExpandableButton: true,
+        maxExpandableHeight: 270,
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/factory/weekView.js]: OK, good point. The visible schedule count will be 10.5 by 230 maximum height.

File: src/js/factory/weekView.js [hunk 7]
```diff
@@ -14,60 +14,80 @@ var Week = require('../view/week/week');
...
-    var weekView, dayNameContainer, dayNameView, vLayoutContainer, vLayout,
-        milestoneView, taskView, alldayView, timeGridView, alldayPanel;
+    var panels = options.week.panels || DEFAULT_PANELS,
+        vpanels = [];
+    var weekView, dayNameContainer, dayNameView, vLayoutContainer, vLayout;
+
+    util.extend(options.week, {panels: panels});
 
-    weekView = new Week(null, options.week, layoutContainer);
+    weekView = new Week(null, options.week, layoutContainer, panels);
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/factory/weekView.js]: I didnt' think about this. I agree with you.
But I will be able to handle it at next refactoring.

File: src/js/handler/daygrid/click.js [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+DayGridClick.prototype._onClick = function(clickEvent) {
+    var self = this,
+        target = clickEvent.target,
+        timeView = this.checkExpectCondition(target),
+        scheduleCollection = this.controller.schedules,
+        collapseElement = domutil.closest(
+            target,
+            config.classname('.weekday-collapse-btn')
+        ),
+        moreElement = domutil.closest(
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/click.js]: I accept that. But little different.
`collapseElement` will be `collapseBtnElement`.
`moreElement` will be `expandBtnElement`.
Isn't it more clear?

File: src/js/factory/weekView.js [hunk 9]
```diff
@@ -76,33 +96,32 @@ module.exports = function(baseController, layoutContainer, dragHandler, options)
...
+            if (vpanels.length) {
+                vpanels.push({
                     isSplitter: true
                 });
             }
-            panels.push(util.extend({
+            vpanels.push(util.extend({
                 name: name
-            }, view));
+            }, panel));
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/factory/weekView.js]: Thanks.

File: src/js/handler/daygrid/click.js [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+ * Check target element is expected condition for activate this plugins.
+ * @param {HTMLElement} target - The element to check
+ * @returns {string} - model id
+ */
+DayGridClick.prototype.checkExpectCondition = DayGridMove.prototype.checkExpectedCondition;
+
+/**
+ * Click event handler
+ * @param {object} clickEvent - click event data
+ * @emits DayGridClick#clickSchedule
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/click.js]: Oh, my god. 
What a helpful code review!

File: src/js/handler/daygrid/core.js [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+/**
+ * @mixin dayGridCore
+ */
+var dayGridCore = {
+    /**
+     * @param {view} view - view instance.
+     * @param {MouseEvent} mouseEvent - mouse schedule object.
+     * @returns {function|boolean} function that return schedule data by mouse events.
+     */
+    _retriveScheduleData: function(view, mouseEvent) {
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/core.js]: There are many same typo error.
We may handle it next time.

File: src/js/handler/daygrid/click.js [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+    this.dragHandler.off(this);
+    this.view = this.controller = this.dragHandler = null;
+};
+
+/**
+ * Check target element is expected condition for activate this plugins.
+ * @param {HTMLElement} target - The element to check
+ * @returns {string} - model id
+ */
+DayGridClick.prototype.checkExpectCondition = DayGridMove.prototype.checkExpectedCondition;
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/click.js]: Yes, you're right. It's `DayGridSchedule` view.

File: src/js/handler/daygrid/click.js [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+/**
+ * Click event handler
+ * @param {object} clickEvent - click event data
+ * @emits DayGridClick#clickSchedule
+ */
+DayGridClick.prototype._onClick = function(clickEvent) {
+    var self = this,
+        target = clickEvent.target,
+        timeView = this.checkExpectCondition(target),
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/click.js]: OK, but i change it to `dayGridScheduleView`.

File: src/js/handler/daygrid/creation.js [hunk 14]
```diff
@@ -0,0 +1,307 @@
...
+    end = datetime.end(dateRange[xIndex]);
+
+    /**
+     * @event {DayGridCreation#beforeCreateSchedule}
+     * @type {object}
+     * @property {string} category - schedule category
+     * @property {boolean} isAllDay - whether schedule is fired in view area?
+     * @property {Date} start - select start time
+     * @property {Date} end - select end time
+     * @property {TimeCreationGuide} guide - TimeCreationGuide instance
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/creation.js]: yes, it it.

File: src/js/handler/daygrid/move.js [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+var dayGridCore = require('./core');
+var DayGridMoveGuide = require('./moveGuide');
+var TZDate = require('../../common/timezone').Date;
+
+/**
+ * @constructor
+ * @implements {Handler}
+ * @mixes dayGridCore
+ * @mixes CustomEvents
+ * @param {Drag} [dragHandler] - Drag handler instance.
```
[the remark this answers is not in the document]
Comment (dongsik-yoo) [reply src/js/handler/daygrid/move.js]: OK. All parameters are required.

File: src/js/handler/daygrid/moveGuide.js (hunk 3 above)
Comment (dongsik-yoo) [reply src/js/handler/daygrid/moveGuide.js]: Thank for your review.

File: src/js/handler/daygrid/resizeGuide.js (hunk 4 above)
Comment (dongsik-yoo) [reply src/js/handler/daygrid/resizeGuide.js]: I changed it.

File: src/js/handler/daygrid/resizeGuide.js (hunk 5 above)
Comment (dongsik-yoo) [reply src/js/handler/daygrid/resizeGuide.js]: Ok. I will handle it at next refactoring.

**Pull request merged**